__app__/crop/ingress_hyper.py

```diff
@@ -0,0 +1,211 @@
+"""
+Python module to import data using the Hyper.ag API
+"""
+
+from collections.abc import Iterable
+import logging
+import time
+from datetime import datetime, timedelta
+import requests
+import pandas as pd
+
+from sqlalchemy import and_
+from sqlalchemy.exc import ProgrammingError
+
+from __app__.crop.db import connect_db, session_open, session_close
+from __app__.crop.structure import (
+    TypeClass,
+    SensorClass,
+    ReadingsAranetTRHClass,
+)
+from __app__.crop.utils import query_result_to_array
+from __app__.crop.constants import CONST_CROP_HYPER_APIKEY, CONST_ARANET_TRH_SENSOR_TYPE
+
+from __app__.crop.ingress import log_upload_event
+from __app__.crop.sensors import get_aranet_trh_sensor_data
+
+CONST_CHECK_URL_PATH = "https://zcf.hyper.ag/api/sites/0/analytics/v3/device_metrics"
+
+
+def get_api_sensor_data(api_key, dt_from, dt_to):
+    """
+    Makes a request to download sensor data for a specified period of time.
+    Note that unlike the equivalent Zensie function, this gets data for _all_ sensors,
+    and returns it as a dict, keyed by Aranet id.
+
+    Arguments:
+        api_key: api key for authentication
+        dt_from: date range from
+        dt_to: date range to
+    Return:
+        success: whether data request was succesful
+        error: error message
+        data_df_dict: dictionary of {aranet_pro_id (str): DataFrame of sensor_data}
+    """
+
+    success = True
+    error = ""
+    data_df_dict = {}
+
+    headers = {
+        "Content-Type": "application/json",
+        "Authorization": "Bearer " + api_key,
+    }
+
+    dt_from_iso = dt_from.strftime("%Y-%m-%dT%H:%M:%S") + "Z"
+    dt_to_iso = dt_to.strftime("%Y-%m-%dT%H:%M:%S") + "Z"
+
+    url = CONST_CHECK_URL_PATH
+
+    params = {
+        "start_time": dt_from_iso,
+        "end_time": dt_to_iso,
+        "metrics": "aranet_temperature,aranet_humidity",
+        "resolution": "10m",
+        "metadata": "true",
+    }
+    response = requests.get(url, headers=headers, params=params, verify=False)
+
+    if response.status_code != 200:
+        error = "Request's [%s] status code: %d" % (url[:70], response.status_code)
+        success = False
+        return success, error, data_df_dict
+    # if we got to here, we have an API response for all T/RH sensors
+    data = response.json()
+    # the metadata contains a mapping between the Hyper id (MAC address) and the
+    # aranet_pro_id
+    device_mapping = data["metadata"]["devices"]
+    timestamps = data["labels"]
+    value_dicts = data["series"]
+    for k, v in device_mapping.items():
```

Comment:
Not a major thing, but I noticed that this for-loop will perform quadratic lookup on `number of devices` * `series (number of devices * number of metrics)`. This is because the for-loop below (L86) performs a linear scan for every device.

Assuming we keep adding more devices and metrics this might unnecessarily slow down processing.

Something like the following does a single linear scan on the series:

```python
    for value_dict in value_dicts:
        hyper_id = value_dict["device_id"]
        aranet_pro_id = device_mapping[hyper_id]["vendor_device_id"]

        metric_name = value_dict["metric"]
        if metric_name == "aranet_ambient_temperature":
            metric_name = "Temperature"
        elif metric_name == "aranet_relative_humidity":
            metric_name = "Humidity"
        
        if aranet_pro_id not in data_df_dict:
            data_df = pd.DataFrame({"Timestamp": pd.to_datetime(timestamps)})
            data_df.set_index("Timestamp", inplace=True)
            data_df_dict[aranet_pro_id] = data_df
        else:
            data_df = data_df_dict[aranet_pro_id]
        data_df[metric_name] = value_dict["values"]
```

The check for timezone information is also not necessary as our API will always include the TZ information (in UTC).

Reply:
I also used the fully qualified names in the renaming as mentioned in my comment above.

Reply:
Good call - thanks again Rizo - making this change now.